### PR TITLE
Add option: --single  to output zero-dependency single html file

### LINKED ISSUE #801: Bundle apidoc file into single html file
Hello guys, I search in the issues and didn't find something similar. Is there a way to serve the apidoc as a single html file and not as a directory? I need this in order to make the apidoc page password protected and I want to serve it with express. Any ideas?

### PR DESCRIPTION
**Pull request by eyasliu**
In many scenarios, it is convenient to output single file instead of folder. I added a cli param `--single` to output a zero-dependency html file

I also provide a shell script (`template-single/build-single.sh`) to generate a single file template based on the original template

it resolve #801 #436

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -52,7 +54,13 @@ function createDoc(options) {
 
     // Paths.
     options.dest     = path.join(options.dest, './');
-    options.template = path.join(options.template, './');
+
+    if (options.single) {
+        options.template = path.join(options.templateSingleFile, './');
+        options.dest = path.join(options.dest, 'docs.html');
```
Comment (NicolasCARPi) [inline lib/index.js]: Maybe call it "index.html" instead of "docs.html"

File: lib/index.js [hunk 2]
```diff
@@ -190,6 +204,24 @@ function createOutputFiles(api) {
         fs.writeFileSync(app.options.dest + './api_definition.js', 'define({ "api": ' + api.definitions + ' });' + '\n');
 }
 
+
+function createSingleFile(api) {
+    if (app.options.simulate)
+        app.log.warn('!!! Simulation !!! No file or dir will be copied or created.');
+
+    // app.log.verbose('create dir: ' + app.options.dest);
```
Comment (NicolasCARPi) [inline lib/index.js]: Remove commented code

File: template-single/build-single.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
Comment (NicolasCARPi) [inline template-single/build-single.sh]: use `#!/usr/bin/env bash`

File: template-single/build-single.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+# build single html file tpl
+
+set -e
+
+SCRIPTPATH=$(cd `dirname $0`; pwd)
```
Comment (NicolasCARPi) [inline template-single/build-single.sh]: use https://www.shellcheck.net/ and fix all issues

File: template-single/build-single.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+node scripts/r.js -o cssIn=scripts/style.css out=tmp/style.css optimizeCss=standard
+
+cp -f scripts/raw.html index.html
+
+echo "<style>$(cat tmp/style.css)</style>" >> index.html
+
+echo "<script>$(cat tmp/main.js)</script>" >> index.html
+echo "<script>$(cat tmp/main.js)</script>" >> index.html
+
+# rm -rf tmp
```
Comment (NicolasCARPi) [inline template-single/build-single.sh]: remove commented code and add a newline at end of file

File: template-single/scripts/api_data.js [hunk 6]
```diff
@@ -0,0 +1 @@
+define(APIDATA)
```
Comment (NicolasCARPi) [inline template-single/scripts/api_data.js]: add newline at end of file

File: template-single/scripts/api_project.js [hunk 7]
```diff
@@ -0,0 +1 @@
+define(APIPROJECT)
```
Comment (NicolasCARPi) [inline template-single/scripts/api_project.js]: add newline at end of file

File: lib/index.js (hunk 2 above)
Comment (eyasliu) [reply lib/index.js]: this is verbose log,I should add it

File: template-single/build-single.sh (hunk 5 above)
Comment (eyasliu) [reply template-single/build-single.sh]: It clean tmp dir, should enable.

**Comment by eyasliu** [comment]
Because it's the most convenient way to run it, but it's not good way.

**Comment by eyasliu** [comment]
If in node_modules, I can only run like this, it doesn't look good

```
npx r.js xxx
```
or

```
node ../node_modules/r.js/xxx.js xxx
```

**Comment by NicolasCARPi** [comment]
In fact requirejs is already in `template/vendor/require.min.js`. I think there is too much duplication in your proposed changes. Maybe keep the `template` folder, and add inside an `index-single.html` if needed, or better, modify the existing index.html and add a switch for the case where it's single and some html needs to be modified.

**Comment by eyasliu** [comment]
> In fact requirejs is already in template/vendor/require.min.js.

`require.js` and `r.js` are different, `r.js` is the nodejs tool. So this file `template/vendor/require.min.js` cannot be used to build pack js and css

> Maybe keep the template folder, and add inside an index-single.html

When generating api files, the entire `template` folder will be copied. If I add any files here, these files will be copied, which is not friendly generated without the `--single` parameter

> modify the existing index.html and add a switch for the case where it's single and some html needs to be modified.

The file `template-single/scripts/raw.html` removes all html tags that reference js and css file, and adds a script tag that sets global variables. Maybe this can be achieved by string replacement based on `template/index.html` file

But without `template-single` folder, where should the other files needed for the build be placed?

**Comment by eyasliu** [comment]
`template-single` folder just two files

* `build-single.sh` is used to generate the `template-single/index.html` file, and other dependent files are temporarily generated
* `index.html` is a template for generating single file documents, the content is generated based on `template/index.html`

**Comment by NicolasCARPi** [comment]
Note to self: close https://github.com/apidoc/apidoc/issues/666 (the devil's issue!!)

File: template-single/build-single.sh [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+# build single html file tpl
+set -e
+
+# fix script cwd
+DIRNAME=$(dirname "$0")
+SCRIPTPATH=$(cd "$DIRNAME"; pwd)
+cd "$SCRIPTPATH"
+
+# define path
+TEMPDIR="tmp"
```
Comment (NicolasCARPi) [inline template-single/build-single.sh]: Please use mktemp. See https://code-maven.com/create-temporary-directory-on-linux-using-bash

File: template-single/index.html [hunk 9]
```diff
@@ -0,0 +1,3271 @@
...
...ve>.table-bordered>tbody>tr>td:first-child,.table-responsive>.table-bordered>tbody>tr>th:first-child,.table-responsive>.table-bordered>tfoot>tr>td:first-child,.table-responsive>.table-bordered>tfoot>tr>th:first-child,.table-responsive>.table-bordered>thead>tr>td:first-child,.table-responsive>.table-bordered>thead>tr>th:first-child{border-left:0}.table-responsive>.table-bordered>tbody>tr>td:last-child,.table-responsive>.table-bordered>tbody>tr>th:last-child,.table-responsive>.table-bordered>tfoot>tr>td:last-child,.table-responsive>.table-bordered>tfoot>tr>th:last-child,.table-responsive>.table-bordered>thead>tr>td:last-child,.table-responsive>.table-bordered>thead>tr>th:last-child{border-right:0}.table-responsive>.table-bordered>tbody>tr:last-child>td,.table-responsive>.table-bordered>tbody>tr:last-child>th,.table-responsive>.table-bordered>tfoot>tr:last-child>td,.table-responsive>.table-bordered>tfoot>tr:last-child>th{border-bottom:0}}fieldset{min-width:0;padding:0;margin:0;border:0}legend{display:block;width:100%;padding:0;margin-bottom:20px;font-size:21px;line-height:inherit;color:#333;border:0;border-bottom:1px solid #e5e5e5}label{display:inline-block;max-width:100%;margin-botto
```
Comment (NicolasCARPi) [inline template-single/index.html]: do we really need to include bootstrap like this?? Surely a better way could be found, no? Even loading it from a CDN would work.

Comment (eyasliu) [reply template-single/index.html]: I think it ’s better to include bootstrap inline, it ’s completely zero-dependency, you may not know that many cdns in China are not accessible

**Comment by eyasliu** [comment]
Thanks for the reply

I fully understand your behavior and I have no negative feelings about it. I have also been making changes in accordance with your suggestions to meet your requirements. There are just a few questions that really need to be discussed with you.

For js and css, I don't advocate using cdn, because in some regions such as China, many cdns are not accessible or unstable, such as unpkg, google, cloudflare

In addition, placing a single-file template in the `template-single` folder will cause the template to fail to take effect when you use the `--template` parameter. Which one do you think is best

**Comment by eyasliu** [comment]
I'm sorry to reply late, because of China's novel coronavirus pneumonia, I'm trapped in the countryside, and I don't have a computer to write code.

**Comment by NicolasCARPi** [comment]
That's a good reason ;) Good luck to you and the ones you love!

**Pull request merged**